File: ChatNestFullStack/ChatNest-app/src/app/features/chat/services/message.service.ts
import { Injectable } from '@angular/core';
import { ApiService } from '../../api.service';
import { Observable } from 'rxjs';
import { DeleteMessageResponseModel, MessagesListResponseModel, SendMessageResponseModel } from 'src/app/models/message-response';

@Injectable({
  providedIn: 'root'
})
export class MessageService {

  constructor(private api: ApiService) { }

  getMessages(chatId: string, userId: string): Observable<MessagesListResponseModel> {
  return this.api.post(`/api/Message/GetMessagesList?chatID=${chatId}&userID=${userId}`, {});
  }

  sendMessage(message: { chatId: string; senderId: string; content: string }): Observable<SendMessageResponseModel> {
    return this.api.post(`/api/Message/SendMessage`, message);
  }


  deleteMessage(messageID: string, userID: string): Observable<DeleteMessageResponseModel> {
    return this.api.post<DeleteMessageResponseModel>(
      `/api/Message/DeleteMessage?messageID=${messageID}&userID=${userID}`,
      {} 
    );
  }

}

File: ChatNestFullStack/ChatNest/Services/MessageService.cs
using ChatNest.Models.Domain;
using ChatNest.Repositories;
using ChatNest.SignalR;
using Microsoft.AspNetCore.SignalR;

namespace ChatNest.Services
{
    public class MessageService : IMessageService
    {
        private readonly IMessageRepository messageRepository;
        private readonly IHubContext<ChatHub> hubContext;

        public MessageService(IMessageRepository messageRepository, IHubContext<ChatHub> hubContext)
        {
            this.messageRepository = messageRepository;
            this.hubContext = hubContext;
        }
        public async Task<DeleteMessageResponseModel> DeleteMessageAsync(Guid messageID, Guid userID)
        {
            return await messageRepository.DeleteMessageAsync(messageID, userID);
        }

        public async Task<MessagesListResponseModel> GetMessagesListAsync(Guid chatID, Guid userID)
        {
            return await messageRepository.GetMessagesAsync(chatID, userID);
        }

        public async Task<MessagesListResponseModel> GetMessagesByIDAsync(Guid requestedMessageID, Guid userID)
        {
            return await messageRepository.GetMessagesByIDAsync(requestedMessageID, userID);
        }

        public async Task<SendMessageResponseModel> SendMessageAsync(Message message)
        {
            var response = await messageRepository.SendMessageAsync(message);

            if (response.MessageID == 1) // success
            {
                // Get the newly created message with enriched data
                var getMessageResponse = await this.GetMessagesByIDAsync(response.NewMessageID, message.senderId);

                var enrichedGroupMessage = getMessageResponse.GroupMessages?.FirstOrDefault();
                var enrichedOneToOneMessage = getMessageResponse.OneToOneMessages?.FirstOrDefault();

                if (enrichedGroupMessage != null)
                {
                    await hubContext.Clients.Group(message.chatId.ToString())
                        .SendAsync("ReceiveMessage", enrichedGroupMessage);
                    Console.WriteLine("It Worked!");
                }
                else if (enrichedOneToOneMessage != null)
                {
                    await hubContext.Clients.Group(message.chatId.ToString())
                        .SendAsync("ReceiveMessage", enrichedOneToOneMessage);
                    Console.WriteLine("It Worked!");
                }
                else
                {
                    // fallback: broadcast minimal payload
                    var broadcast = new GroupMessageResponse
                    {
                        MessageID = response.NewMessageID,
                        ChatID = message.chatId,
                        SenderID = message.senderId,
                        Content = message.content,
                        SentAt = DateTime.UtcNow,
                        DisplayName = "Unknown"
                    };

                    await hubContext.Clients.Group(message.chatId.ToString())
                        .SendAsync("ReceiveMessage", broadcast);
                }
            }

            return response;
        }
    }
}
